### PR TITLE
Added a state property for the Request class

### PR DESCRIPTION
**Pull request by FirstKlaas**
Added a `state` property for the `responder.models.Request` class. The property just exposes the starlette state property. No extra logic added. No testcase written. Added a short docstring to the property.

### CODE REVIEW

**Comment by taoufik07** [comment]
Thanks, can you add some simple tests ?

**Comment by FirstKlaas** [comment]
Ok. Had to read a quick pytest tutorial first. Never used it before. Seems pretty straight forward. Why do you have no makefile and requirements file (for development), no linting... Did I miss something?

**Comment by iancleary** [comment]
This project uses `black` for linting: https://github.com/taoufik07/responder/blob/master/.travis.yml#L17 and pipenv for virtual environments and requirements.
  
The other thing I have found is app.py and app2.py are in the .gitignore file as well https://github.com/taoufik07/responder/blob/master/.gitignore#L16-L17 in case you need

https://python-responder.org/en/latest/testing.html (possibly could use some updates per this process)

Not sure if I missed anything, @taoufik07 ?

**Comment by taoufik07** [comment]
in addition to what @iancleary said, we're using `pipenv` so our requirements file is the Pipfile and Pipfile.lock.

For dev purposes you need to use `pipenv install --dev` in order to install dev dependencies or if you don't use `pipenv` go to the Pipfile grab the packages names and install them.

@FirstKlaas if you have some hard time writing the tests I can assist you

**Comment by taoufik07** [comment]
The test can be something like 

```python
def test_request_state(api): 
     @api.route("/")      
     async def home(req, resp): 
         req.state.username = "Klaas" 
         resp.text = "First {}".format(req.state.username) 

     r = api.requests.get("/")
     assert r.text == "First Klaas"
```

**Comment by FirstKlaas** [comment]
Hi, the "pipenv" info was the point I was missing. Interesting. I added a test case. Thanks for the answers to my question. So I think

**Comment by FirstKlaas** [comment]
Great @taoufik07. My test is slightly different, as I added a minimal middelware. I think this is more realistic.

File: tests/test_request_state.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+from starlette.middleware.base import BaseHTTPMiddleware
+
+
+class StateMiddleware(BaseHTTPMiddleware):
+
+    async def dispatch(self, request, call_next):
+        request.state.test1 = int(12)
```
Comment (taoufik07) [inline tests/test_request_state.py]: no need to call the int

File: tests/test_request_state.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        request.state.test2 = "Foo"
+
+        response = await call_next(request)
+        return response
+
+
+def test_api_request_state(api, url):
+
```
Comment (taoufik07) [inline tests/test_request_state.py]: can you move the middleware inside the test function

File: tests/test_request_state.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    
+    api.add_middleware(StateMiddleware)
+
+    @api.route("/")
+    def home(req, resp):
+        assert hasattr(req, "state")
+        assert hasattr(req.state, "test1")
+        assert hasattr(req.state, "test2")
+        assert isinstance(req.state.test1, int)
+        assert isinstance(req.state.test2, str)
```
Comment (taoufik07) [inline tests/test_request_state.py]: `assert request.state.test1 == 12` and `assert request.state.test2 == "foo"` are enough no need for the `hasattr` or `isinstance`

Comment (taoufik07) [reply tests/test_request_state.py]: I don't think having the tests inside the home view is a good idea, you can simply return the state values in `resp.text` and do the test

File: Pipfile [hunk 4]
```diff
@@ -15,6 +15,7 @@ flask = "*"
 sphinx = "*"
 marshmallow = "*"
 pytest-cov = "*"
+pylint = "*"
```
Comment (taoufik07) [inline Pipfile]: can you please remove `pylint` ?

File: Pipfile [hunk 5]
```diff
@@ -4,7 +4,7 @@ verify_ssl = true
 name = "pypi"
 
 [packages]
-responder = {editable = true, path = "."}
+responder = {editable = true,path = "."}
```
Comment (taoufik07) [inline Pipfile]: and add the space after the comma

**Comment by FirstKlaas** [comment]
Ok. I removed the asserts from the home view. Although I think it was a good idea. Just to make shure, that no type conversion happen during chaining. But I removed them anyway. I ran black. Now everything should be fine. I'm happy when this will be released, as I'm working on a larger plattform and decided to go with responder. Great library so far. Keep up the good work.

**Comment by FirstKlaas** [comment]
Again sorry. Overlooked your request to move the test. Wait a second.

**Comment by taoufik07** [comment]
@FirstKlaas thanks for contribution !

**Comment by taoufik07** [comment]
Can you squash the commits to keep the log tree clean before I merge ?

**Comment by taoufik07** [comment]
You can use something like `git reset --soft HEAD~6 && git commit -m "Add state for requests"`

**Comment by taoufik07** [comment]
Just try with `git reset --soft HEAD~8 && git commit -m "Add state for requests"` and then force push `git push -f origin iss382-state-for-request` so we end up with a clean reflog

**Comment by FirstKlaas** [comment]
Now i ruined everything. :( Now it says Pipfile.lock is conflicted. But locally I have no conflicts. I just copied the two git commands you wrote, as I don't fully understand, what they are doing. I would be easier to create a new branch, do the clean changes and then create a new pull request. Otherwise I have no clue what to do.

**Comment by taoufik07** [comment]
@FirstKlaas ok xD

**Comment by FirstKlaas** [comment]
Hi @taoufik07, created a new branch with a clean implementation, testcase and documentation. Created a new pull request based on the new branch. I will delete this branch and you can fire and forget this pull request. Thanks for your patience and your support. I learned two new things, which is great. pipenv and pytest. The latter I knew, but never used. The first was totally new to my. Next contribution will be smoother. ;)

**Comment by taoufik07** [comment]
closed via #385

**Pull request closed without merging**